FLDyn outlet process bug correction

rubenzorrilla opened the pull request.

rubenzorrilla committed:
Barrier() exported to Python

rubenzorrilla committed:
Bug correction in average outlet velocity computation

rubenzorrilla committed:
Indentation mistake

RiccardoRossi commented on applications/FluidDynamicsApplication/python_scripts/apply_outlet_process.py:
```diff
@@ -128,15 +128,17 @@ def _ComputeOutletCharacteristicVelocity(self):
         for node in self.outlet_model_part.GetCommunicator().LocalMesh().Nodes:
             vnode = node.GetSolutionStepValue(KratosMultiphysics.VELOCITY)
             outlet_avg_vel_norm += math.sqrt(vnode[0]*vnode[0] + vnode[1]*vnode[1] + vnode[2]*vnode[2])
-            outlet_avg_vel_norm /= len(self.outlet_model_part.GetCommunicator().LocalMesh().Nodes)
```
i'll accept this for now, however it should be definitely be implemented in c++ (it is a good candidate for a function in variable utils)

RiccardoRossi replied:
take care, in mpi you need to sync to min...

RiccardoRossi commented on applications/FluidDynamicsApplication/python_scripts/apply_outlet_process.py:
```diff
@@ -128,15 +128,17 @@ def _ComputeOutletCharacteristicVelocity(self):
+        outlet_avg_vel_norm /= len(self.outlet_model_part.GetCommunicator().LocalMesh().Nodes)  # Partition average value
+
+        # Perform MPI communication
+        self.outlet_model_part.GetCommunicator().Barrier()                                      # Wait for all the processors to compute the average
```
SumAll should have an implicit barrier. I don't think you need the barrier before

rubenzorrilla replied:
Ok!! In any case, shall we keep the Barrier() method exported to Python?

RiccardoRossi commented on applications/FluidDynamicsApplication/python_scripts/apply_outlet_process.py:
```diff
@@ -128,15 +128,17 @@ def _ComputeOutletCharacteristicVelocity(self):
-            else:
-                self.outlet_model_part.GetRootModelPart().ProcessInfo[KratosFluid.CHARACTERISTIC_VELOCITY] = min_outlet_avg_vel_norm
+
+        outlet_avg_vel_norm /= len(self.outlet_model_part.GetCommunicator().LocalMesh().Nodes)  # Partition average value
```
this is not correct. You need to get the global sum, and the global number of nodes in the model part... NOT the local one. otherwise the result will not be the same of all the processors

rubenzorrilla replied:
What I'm trying to do is to compute the local average value in each processor (which is not the same in all processors as you point). Below I compute the average of the local averages to get the average between processors dividing the result of the SumAll by the number of procs. Is this not correct?

RiccardoRossi reviewed (changes requested):
i think there are some parallelization errors to be ironed out still...

rubenzorrilla committed:
MPI correction

Correction in the outlet average velocity computation. Now it is ensured that the average velocity is the same in all processes.

RiccardoRossi reviewed (approved):
ok,
it now looks ok to me
